### DISCUSSION TITLE
500 Error for /API/internalservices/controlBar/ToggleUserMode after 2sxc Upgrade

### DISCUSSION THREAD

**Discussion by Chris-Behling** [2sic/2sxc#3432] [Q&A]
Hey all, I have a site I'm working on some upgrades. 

- The production site is on DNN 9.10.02 and 2sxc 14.7.3
- Upgrading DNN to 9.13.03, and everything works fine.
- At some point, upgrading 2sxc causes 500 error when trying to get into DNN edit mode. 

We've tried:
- Upgrading straight to 2sxc 17.09, and it produces this error.
- Upgrading in steps. 14.07.04 (works), 14.12.04 (works), 16.00.00 (causes error).

In the stepped approach, I've also installed other, more recent versions of 2sxc, but it still produces this error. 
The install works without throwing any error, it just seems to cause the 500 error when trying to get into edit mode.

Any ideas or recommendations on this issue and how to get past it?

**Comment by iJungleboy** [comment]
I remember having seen this before. Can't recall the details but it had something to do with services missing (Dependency Injection) which were accessed on all API calls. 

You should see more information about this in the DNN Logs - I belive you'll see errors and more. Best check that and add below.

**Comment by iJungleboy** [comment]
Note: I found a Discord thread from January that reported the same issue, maybe he found out something then

Not sure if this link will work https://discord.com/channels/706654343176126496/1037696933805965382/1201975254566830162

**Reply by Chris-Behling** [reply]
@iJungleboy, that link doesn't work for me. Could you share the name of the server (or an invite) and the date/time of the post so I can track it down? My username on Discord is 10poundgorilla_78297, or a screenshot of the post might suffice.

**Reply by iJungleboy** [reply]
The discord is linked here https://dnncommunity.org/

**Reply by trouble2** [reply]
Hi there, I've just ran into this issue again on 2sxc 18.6.1 on dnn 9.11.2 then tried upgading to dnn 9.13.3 and reinstalling 2sxc.
That didn't help. 
Can't seem to find the solutions here, but apparently you know more, so can you post the solution?

**Reply by trouble2** [reply]
What I can see in the log file is:
[ERROR] DotNetNuke.Common.Internal.EventHandlersContainer`1[[DotNetNuke.Entities.Portals.IPortalEventHandlers, DotNetNuke, Version=9.13.3.0, Culture=neutral, PublicKeyToken=null]] - De type-initialisatiefunctie voor DotNetNuke.ExtensionPoints.ExtensionPointManager heeft een uitzondering veroorzaakt.
System.TypeInitializationException: De type-initialisatiefunctie voor DotNetNuke.ExtensionPoints.ExtensionPointManager heeft een uitzondering veroorzaakt. ---> System.TypeLoadException: Kan type ToSic.Eav.Documentation.PublicApi niet laden vanuit assembly ToSic.Eav.Core, Version=18.6.1.0, Culture=neutral, PublicKeyToken=null.
   bij System.ModuleHandle.ResolveType(RuntimeModule module, Int32 typeToken, IntPtr* typeInstArgs, Int32 typeInstCount, IntPtr* methodInstArgs, Int32 methodInstCount, ObjectHandleOnStack type)
   bij System.ModuleHandle.ResolveTypeHandleInternal(RuntimeModule module, Int32 typeToken, RuntimeTypeHandle[] typeInstantiationContext, RuntimeTypeHandle[] methodInstantiationContext)

**Reply by trouble2** [reply]
So, it looks like some upgrades go ok, and some have this issue.
If you have information what the cause is, or how I can prevent it, please let me know.
Right now, I cannot upgrade this client because after the upgrade we cannot edit anymore

**Reply by iJungleboy** [reply]
@trouble2 You must have some old 2sxc DLL in the system, one that is not included in any new distribution (or wasn't correctly replaced).

Reason: the class `ToSic.Eav.Documentation.PublicApi` was moved to `ToSic.Lib.Documentation.PublicApi` a LONG time ago. ca. December 2022 https://github.com/2sic/eav-server/commits/master/ToSic.Lib.Core/Documentation/PublicApi.cs

It is used in various places to mark APIs so the docs know what to publish, or what is internal.

So the fact that your system seems to want this old name means that you must have some DLL which is loaded that's at least 2 years old and should not be in use.

**Reply by trouble2** [reply]
Aha, ok.
It was an upgrade from 13.3 to 18.6 or something like that...
So this is not deleted in the upgrade for some reason?
Is there an upgrade path, or a way to check which files to delete?
Looks like I should delete ToSic.Eav.Tokens.dll
Maybe you should check for that file in the upgrade script somewhere?

**Reply by iJungleboy** [reply]
I don't know what happened on your side, because we do clean up on upgrades.
Maybe the file was even older.

Basically after a 2sxc upgrade all 2sxc files should have a fairly recent date, except maybe some bundled DLLs like RazorBlade or Koi, which can be older as these libraries don't get recompiled
